### PR TITLE
Adjacency

### PR DESCRIPTION
**Pull request by Werdna6001**

### CODE REVIEW

**Comment by Werdna6001** [comment]
I disagree with the error that the tests are finding. 
It doesn't seem to be finding the Board_test.py file.

File: src/Tile.py
Comment (iota-pi) [inline src/Tile.py]: I think this and the below additions of `self.` are not correct. Perhaps to clarify, it would be good to call the parameter `new_building`?

Comment (iota-pi) [inline src/Tile.py]: The issue the tests are picking up here is actually correct. I forgot to initialise the value of `self.building = None` in `__init__`

File: src/Tile_test.py
Comment (iota-pi) [inline src/Tile_test.py]: Can I suggest adding `not spot.can_build(Building.TRADING_HOUSE)` here too and the similar cases below? (I see that `Building.DWELLING` has it above)

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: Could you please use `snake_case` for variable and function names? No particular reason aside from that's just the standard practice in Python. `CamelCase` for classes though

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: This check gets repeated a lot. How about we pull it to the bottom of the function. Something like:
```python
from typing import Sequence, List

def get_directly_adjacent(..., terrain_filter: Sequence[Terrain] | None) -> List[tuple[int, int]]:
    ...

    filtered_terrain = [(x, y) for x, y in adjacent_tile_list if self.get(x, y).terrain not in terrain_filter]]
```

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: This could be more efficient by using a Set. That should be able to make it O(n) instead of O(n^2). Have a `found` set that you add elements to as well as the `new_list` variable and use the set to check if the item has been found already.

**Review by iota-pi — COMMENTED** [review]
Thanks for this! I haven't had a chance to finish looking through it yet but hopefully these are helpful comments to get started on 🙂

File: src/Board_test.py
Comment (iota-pi) [inline src/Board_test.py]: A late comment, but just realised why this file doesn't get run by pytest.. it wants the name to start with `Test` instead of finish with it, so it would need to be `TestBoard`

**Comment by Werdna6001** [comment]
I think this is an error in the test file and that the code is actually fine.

**Comment by Werdna6001** [comment]
I had tried to define 2 Board.get() functions and have them differentiated by their parameters. apparently that's just a java thing.

**Comment by Werdna6001** [comment]
this last fails 2 tests, the "invalid number of arguments" error is wrong. (I am not giving it 3 arguments.)
The other test it fails is that I was testing for a negative case, (Auren can't build on mountains) and instead of returning false and passing the test it raised an error and failed the test. I think we should replace: 

            raise InvalidActionError(
                f"Faction {faction.name} cannot build on {self._terrain.name}"
            )
and
            raise InvalidActionError(
                "Cannot build on top of another faction's building"
            )
with
            return False

I didn't change it though because you put that there and may have had a reason for it.

**Comment by iota-pi** [comment]
> I had tried to define 2 Board.get() functions and have them differentiated by their parameters. apparently that's just a java thing.

Ah yes, in Python, it's normal to just define one method and handle the different possible types within the function.. or often it's just preferable to have different names for the functions

File: src/Tile_test.py
Comment (iota-pi) [inline src/Tile_test.py]: This is a common thing to do in Java but is generally not preferred in Python because it makes dependencies less clear, has a performance overhead when this module is first loaded, and clutters the namespace (i.e. imports a bunch of extra variable names)

So even though it's quite verbose, I'd prefer listing these all out:
```suggestion
from Faction import (
    Dwarves,
    Engineers,
    Auren,
    Witches,
    Swarmlings,
    Mermaids,
    Alchemists,
    Darklings,
    Halflings,
    Cultists,
    Fakirs,
    Nomads,
    ChaosMagicians,
    Giants,
)
```

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: It would be good to have typing information for this function (P.S. do you have `strict` mode enabled in your vscode? It should complain about missing typing info here if so).

I'd suggest defining a type early in the file:
```python3
Coords = Tuple[int, int]
```

```suggestion
    def get(self, coords: Coords):
```
(this would require importing `Tuple` from `typing` too)

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]:
```suggestion
    def get_directly_adj(
        self,
        start: Coords = (0, 0),
        terrain_filter: Terrain | None = None,
    ) -> List[Coords]:
```

Mostly adding typing but also `terain` -> `terrain`

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: Instead of indenting this here and then re-calling the function in the `else` block, what if we just check at the end (where the corresponding `else` is)

```python3
if terrain_filter is not None:
    adjacent_tile_list = [coord for coord in adjacent_tile_list if self.get(coord)._terrain == terrain_filter]
```

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: This is not really important but it occurs to me that we could make this more concise if, at the start, we define:
```python3
row_start_offset = -(start[1] % 2)
```

And then this could just be
```suggestion
                adjacent_tile_list.append((start[0] + row_start_offset , start[1] - 1))  # top left
                adjacent_tile_list.append((start[0] + row_start_offset + 1, start[1] - 1))  # top right
```

And similarly for the bottom row

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: Similar typing and `terrain_filter` spelling fix would be needed here

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: I don't think we need a separate function for this.. in Python if you're comparing tuples like this you can just do:
```suggestion
                if tile not in adjacent_tile_list:
```

File: src/Board.py
Comment (iota-pi) [inline src/Board.py]: Took me a while to understand the test failures when calling this function, but the issue stems from how Python handles class methods.
```python3
b = Board(...)

# This:
b.found_tile([], (1,2))
# is equivalent to:
Board.found_tile(b, [], (1,2))
```
So because found_tile is an instance method, Python will pass the class instance as the first parameter automatically (which is why it then complains that there are 3).
To get around this, you could either define this as a static method by adding the `@staticmethod` decorator on the line above the function definition or add the `self` parameter here: `def found_tile(self, found, ...):`

But all of that is more for interest's sake than anything else because, as I mentioned in another comment, we should be able to get away with just using the `in` operator in Python... isn't it so much nicer than Java? 😉

File: src/Board_test.py
Comment (iota-pi) [inline src/Board_test.py]: Sorry, I haven't looked through all of the history to know why you've commented out these test cases, but I'd prefer if they're included if they're working, fixed if they're not, or removed entirely if they're not needed (otherwise in a few weeks/months time we'll have no idea what the original plan was for them)

**Review by iota-pi — COMMENTED** [review]
Thanks for working on this! I think the code logic itself is good, these are just a few suggestions about code clarity and the like. I've also made a few changes on this branch to fix a typing issue that was my fault and to make the change to the `can_build` function that you suggested

File: src/Board.py
Comment (Werdna6001) [reply src/Board.py]: I thought that I did have the self parameter in there.

File: src/Board_test.py
Comment (Werdna6001) [reply src/Board_test.py]: I wasn't sure that the answer they were checking for was correct, I had planned to get the tests working without them then add them back in one by one, until they were all back.

**Comment by Werdna6001** [comment]
Earlier today I couldn't work out how to reopen vscode in the container, but I could work out how to run the formatter. Now I've managed to open vscode in the dev container, but I can't get the formatter to run.
I keep force pushing because I've amended the last commit so as not to have hundreds of commits by the time we're trying to merge it.

would it be possible to check format after running the tests? I just had to re run the whole thing because of 2 spaces that weren't "good formatting" I'm actually trying to see if the code works, not if it looks marginally nicer.

**Comment by iota-pi** [comment]
> Earlier today I couldn't work out how to reopen vscode in the container, but I could work out how to run the formatter. Now I've managed to open vscode in the dev container, but I can't get the formatter to run. I keep force pushing because I've amended the last commit so as not to have hundreds of commits by the time we're trying to merge it.

No worries. You're welcome to amend the commits if you like but I also don't mind having lots of commits. On the weekend, if you're interested, I can also show you an easy way to "squash" commits together before merging

> would it be possible to check format after running the tests? I just had to re run the whole thing because of 2 spaces that weren't "good formatting" I'm actually trying to see if the code works, not if it looks marginally nicer.

Yes, good plan. I'll put in a PR for it now.

**Comment by Werdna6001** [comment]
Don't merge this yet, I haven't added provision for if you're the mermaids yet. (the mermaids can count one river tile as directly adjacent) or do you think that calculation should be in the mermaids class?

File: src/Board.py
Comment (iota-pi) [inline src/Board.py:123]: As a general note, instead of having "magic numbers" (or magic strings in this case) in the middle of code, it's better to have some kind of a value that can be type-checked and have other static analysis (i.e. analysis that doesn't involve running the code) run on it.

For example, this could use an Enum:
```python3
from enum import Enum

class AdjacencyType(Enum):
    DIRECT = 0
    INDIRECT = 1

...

return AdjacencyType.DIRECT
```

**Review by iota-pi — APPROVED** [review]
Great, thanks! I've made one suggestion but I'm happy to merge this and we'll fix it later

**Pull request merged**
